### PR TITLE
Binary operations and tests in Numbers: TInt and TFloat

### PR DESCRIPTION
**Pull request by ricknjacky**
Issue reference:- #6

### CODE REVIEW

**Comment by ricknjacky** [comment]
I have just added one test for the binary operations added, @meganindya if this is what we want, 
I will add the tests for the rest, pls provide feedback as usual, Thanks

File: src/.prototype/syntaxElements/primitiveElements.test.ts [hunk 1]
```diff
@@ -194,6 +194,9 @@ describe('operations on type elements', () => {
         test('mod TInt(15) and TInt(4) and expect new element value to be 3', () => {
             expect(TInt.mod(operand_1, operand_2).value).toBe(3);
         });
+        test('check for equality TInt(15) and TInt(4) and expect to get True or False', () => {
```
Comment (meganindya) [inline src/.prototype/syntaxElements/primitiveElements.test.ts:197]: Leave a blank line between tests ... just for readability.

**Review by meganindya — COMMENTED** [review]
Yeah, add positive and negative tests for `equals`, `greaterThan`, `lessThan`. If you can think of any more primitive integer operations, add the methods and exhaustive test cases for them. Commit only when all tests pass.

File: src/.prototype/syntaxElements/primitiveElements.test.ts (hunk 1 above)
Comment (ricknjacky) [reply src/.prototype/syntaxElements/primitiveElements.test.ts:197]: Will do

**Comment by ricknjacky** [comment]
> Yeah, add positive and negative tests for `equals`, `greaterThan`, `lessThan`. If you can think of any more primitive integer operations, add the methods and exhaustive test cases for them.

Done

>Commit only when all tests pass.


Done:-
Screenshot (536)

**Comment by ricknjacky** [comment]
Screenshot (537)


Added operations and tests for **`TFloat`** as well. These were the operations that occurred to me as essential, If we expect more such operations that I have missed, please do tell. I will add them and tests for the same as well. 

PS:- As is, this PR just focuses on **`TFloat`** and **`TInt`**

File: src/.prototype/syntaxElements/primitiveElements.ts [hunk 2]
```diff
@@ -113,6 +113,30 @@ export class TInt extends PrimitiveElement<number> {
     static mod(operand_1: TInt, operand_2: TInt) {
         return new TInt(operand_1.value % operand_2.value);
     }
+        /** Binary == operator. */
```
Comment (meganindya) [inline src/.prototype/syntaxElements/primitiveElements.ts:116]: newline

File: src/.prototype/syntaxElements/primitiveElements.test.ts [hunk 3]
```diff
@@ -256,4 +296,4 @@ describe('operations on type elements', () => {
             expect(TBoolean.or(operand_5, operand_5).value).toBe(false);
         });
     });
-});
+});
```
Comment (meganindya) [inline src/.prototype/syntaxElements/primitiveElements.test.ts:299]: newline

File: src/.prototype/syntaxElements/primitiveElements.ts [hunk 4]
```diff
@@ -287,4 +321,4 @@ export class TBoolean extends PrimitiveElement<boolean> {
     static or(operand_1: TBoolean, operand_2: TBoolean) {
         return new TBoolean(operand_1.value || operand_2.value);
     }
-}
+}
```
Comment (meganindya) [inline src/.prototype/syntaxElements/primitiveElements.ts:324]: newline

File: src/.prototype/syntaxElements/primitiveElements.ts [hunk 5]
```diff
@@ -113,6 +113,30 @@ export class TInt extends PrimitiveElement<number> {
...
+
+    /** Binary ++ operator. */
+    static increment(operand_1: TInt) {
+        return new TInt(operand_1.value + 1)
+    }
+
+    /** Binary -- operator. */
+    static decrement(operand_1: TInt) {
+        return new TInt(operand_1.value - 1)
+    }
```
Comment (meganindya) [inline src/.prototype/syntaxElements/primitiveElements.ts:139]: Unary operators.

Unary operators wouldn't return a new primitive object, instead they'll update the current object.

e.g.
```typescript
increment() {
    (this.value as number)++;
}
```

File: src/.prototype/syntaxElements/primitiveElements.ts [hunk 6]
```diff
@@ -185,6 +209,16 @@ export class TFloat extends PrimitiveElement<number> {
...
+
+    /** Binary ++ operator. */
+    static increment(operand_1: TInt | TFloat) {
+        return new TFloat(operand_1.value + 1)
+    }
+
+    /** Binary -- operator. */
+    static decrement(operand_1: TInt | TFloat) {
+        return new TFloat(operand_1.value - 1)
+    }
```
Comment (meganindya) [inline src/.prototype/syntaxElements/primitiveElements.ts:221]: same as TInt

File: src/.prototype/syntaxElements/primitiveElements.test.ts [hunk 7]
```diff
@@ -218,6 +238,26 @@ describe('operations on type elements', () => {
         test('mod TFloat(15.5) and TInt(4) and expect new element value to be 3.5', () => {
             expect(TFloat.mod(operand_3, operand_2).value).toBe(3.5);
         });
+
+        test('check for equality TFloat(15.5) and TInt(4) and expect to get True or False', () => {
```
Comment (meganindya) [inline src/.prototype/syntaxElements/primitiveElements.test.ts:242]: `'check for equality of TFloat(15.5) and TInt(4) and expect to get false'`

File: src/.prototype/syntaxElements/primitiveElements.test.ts [hunk 8]
```diff
@@ -218,6 +238,26 @@ describe('operations on type elements', () => {
...
+            expect(TFloat.equals(operand_3, operand_2).value).toBe(false);
+        });
+
```
Comment (meganindya) [inline src/.prototype/syntaxElements/primitiveElements.test.ts:245]: add equality positive test

File: src/.prototype/syntaxElements/primitiveElements.test.ts [hunk 9]
```diff
@@ -218,6 +238,26 @@ describe('operations on type elements', () => {
...
+        test('check for equality TInt(15) and TInt(4) and expect to get True or False', () => {
```
Comment (meganindya) [inline src/.prototype/syntaxElements/primitiveElements.test.ts:246]: `'check if TFloat(15.5) greaterThan TInt(4) and expect to get true'`

File: src/.prototype/syntaxElements/primitiveElements.test.ts [hunk 10]
```diff
@@ -218,6 +238,26 @@ describe('operations on type elements', () => {
...
+            expect(TFloat.greaterThan(operand_3, operand_2).value).toBe(true);
+        });
+
+        test('check for equality TInt(15) and TInt(4) and expect to get True or False', () => {
```
Comment (meganindya) [inline src/.prototype/syntaxElements/primitiveElements.test.ts:250]: `'check if TFloat(15.5) lessThan TInt(4) and expect to get false'`

File: src/.prototype/syntaxElements/primitiveElements.test.ts [hunk 11]
```diff
@@ -218,6 +238,26 @@ describe('operations on type elements', () => {
...
+            expect(TFloat.lessThan(operand_1, operand_2).value).toBe(false);
+        });
+
+        test('check if TInt(15) is incremented to be 16.5', () => {
+            expect(TFloat.increment(operand_3).value).toBe(16.5);
+        });
+
+        test('check if TInt(15) is decremented to be 14', () => {
+            expect(TFloat.decrement(operand_3).value).toBe(14.5);
+        });
```
Comment (meganindya) [inline src/.prototype/syntaxElements/primitiveElements.test.ts:260]: Fix accordingly after changes to actual classes.

File: src/.prototype/syntaxElements/primitiveElements.test.ts [hunk 12]
```diff
@@ -194,6 +194,26 @@ describe('operations on type elements', () => {
...
+            expect(TInt.lessThan(operand_1, operand_2).value).toBe(false);
+        });
+
+        test('check if TInt(15) is incremented to be 16', () => {
+            expect(TInt.increment(operand_1).value).toBe(16);
+        });
+
+        test('check if TInt(15) is decremented to be 14', () => {
+            expect(TInt.decrement(operand_1).value).toBe(14);
+        });
```
Comment (meganindya) [inline src/.prototype/syntaxElements/primitiveElements.test.ts:216]: refer comments on TFloat test cases

**Review by meganindya — CHANGES_REQUESTED** [review]

File: src/.prototype/syntaxElements/primitiveElements.ts (hunk 5 above)
Comment (ricknjacky) [reply src/.prototype/syntaxElements/primitiveElements.ts:139]: Thanks used this syntax.
```
increment() {
    ((this.value +1) as number);
}
```

reason:- [link](https://dev.to/theodesp/comment/799h)

**Comment by ricknjacky** [comment]
Thanks for the feedback, have made the changes.

**Comment by ricknjacky** [comment]
### Test log:-
Screenshot (541)

**Comment by meganindya** [comment]
Add non-static `add`, `subtract`, `multiply`, `divide`, `mod` for `TFloat`. I've added for `TInt`. For `TFloat`, the single argument can be of either `TInt` or `TFloat` (`operand: TInt | TFloat`).

Thereafter, add the tests for all 5 + 5 of these methods.

**Comment by ricknjacky** [comment]
Screenshot (545)

@meganindya 
For non-static methods, Are these tests written correctly ?

**Comment by meganindya** [comment]
Yeah. Fine.

File: src/.prototype/syntaxElements/primitiveElements.ts [hunk 13]
```diff
@@ -88,31 +88,77 @@ export class TInt extends PrimitiveElement<number> {
...
+    subtract(operand: TInt) {
+        this.value -= operand.value;
+    }
+
     /** Binary - operator. */
     static subtract(operand_1: TInt, operand_2: TInt) {
         return new TInt(operand_1.value - operand_2.value);
     }
 
+    multiplication(operand: TInt) {
```
Comment (meganindya) [inline src/.prototype/syntaxElements/primitiveElements.ts:111]: `multiply`

File: src/.prototype/syntaxElements/primitiveElements.ts [hunk 14]
```diff
@@ -88,31 +88,77 @@ export class TInt extends PrimitiveElement<number> {
...
+        this.value *= operand.value;
+    }
+
     /** Binary * operator. */
     static multiply(operand_1: TInt, operand_2: TInt) {
         return new TInt(operand_1.value * operand_2.value);
     }
 
+    division(operand: TInt) {
```
Comment (meganindya) [inline src/.prototype/syntaxElements/primitiveElements.ts:120]: `divide`

File: src/.prototype/syntaxElements/primitiveElements.ts [hunk 15]
```diff
@@ -145,32 +191,53 @@ export class TFloat extends PrimitiveElement<number> {
...
+    subtract(operand: TInt | TFloat) {
+        this.value -= operand.value;
+    }
+
     /** Binary * operator. */
     static multiply(operand_1: TFloat | TInt, operand_2: TFloat | TInt) {
         return new TFloat(operand_1.value * operand_2.value);
     }
 
+    multiplication(operand: TInt | TFloat) {
```
Comment (meganindya) [inline src/.prototype/syntaxElements/primitiveElements.ts:219]: `multiply`

File: src/.prototype/syntaxElements/primitiveElements.ts [hunk 16]
```diff
@@ -145,32 +191,53 @@ export class TFloat extends PrimitiveElement<number> {
...
+        this.value *= operand.value;
+    }
+
     /** Binary / operator. */
     static divide(operand_1: TFloat | TInt, operand_2: TFloat | TInt) {
         return new TFloat(operand_1.value / operand_2.value);
     }
 
+    division(operand: TInt | TFloat) {
```
Comment (meganindya) [inline src/.prototype/syntaxElements/primitiveElements.ts:228]: `divide`

**Review by meganindya — CHANGES_REQUESTED** [review]
Put the names back. Function/method names are meant to be imperative by convention.

**Comment by ricknjacky** [comment]
> 
> 
> Put the names back. Function/method names are meant to be imperative by convention.


My bad, did a typo in these two cases ;-; will rebase and push asap.

**Comment by ricknjacky** [comment]
**WIP** The updates values seem to generate errors here, I am using different variables for testing the older tests.

**Comment by ricknjacky** [comment]
@meganindya variables display erratic behavior can you please tell me how do I troubleshoot this?

**Comment by ricknjacky** [comment]
> @meganindya variables display erratic behavior can you please tell me how do I troubleshoot this?

Nvm, I found a way around.

**Comment by ricknjacky** [comment]
Tests

All test cases pass.

**Review by meganindya — APPROVED** [review]

**Pull request merged**